crates/goose/src/session/session_manager.rs

```diff
@@ -0,0 +1,928 @@
+            accumulated_input_tokens: None,
+            accumulated_output_tokens: None,
+            schedule_id: None,
+            recipe: None,
+        }
+    }
+
+    pub fn description(mut self, description: impl Into<String>) -> Self {
+        self.description = Some(description.into());
+        self
+    }
+
+    pub fn working_dir(mut self, working_dir: PathBuf) -> Self {
+        self.working_dir = Some(working_dir);
+        self
+    }
+
+    pub fn extension_data(mut self, data: ExtensionData) -> Self {
+        self.extension_data = Some(data);
+        self
+    }
+
+    pub fn total_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.total_tokens = Some(tokens);
+        self
+    }
+
+    pub fn input_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.input_tokens = Some(tokens);
+        self
+    }
+
+    pub fn output_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.output_tokens = Some(tokens);
+        self
+    }
+
+    pub fn accumulated_total_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.accumulated_total_tokens = Some(tokens);
+        self
+    }
+
+    pub fn accumulated_input_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.accumulated_input_tokens = Some(tokens);
+        self
+    }
+
+    pub fn accumulated_output_tokens(mut self, tokens: Option<i32>) -> Self {
+        self.accumulated_output_tokens = Some(tokens);
+        self
+    }
+
+    pub fn schedule_id(mut self, schedule_id: Option<String>) -> Self {
+        self.schedule_id = Some(schedule_id);
+        self
+    }
+
+    pub fn recipe(mut self, recipe: Option<Recipe>) -> Self {
+        self.recipe = Some(recipe);
+        self
+    }
+
+    pub async fn apply(self) -> Result<()> {
+        SessionManager::apply_update(self).await
+    }
+}
+
+pub struct SessionManager;
+
+impl SessionManager {
+    async fn instance() -> Result<Arc<SessionStorage>> {
+        match SESSION_STORAGE.get() {
+            Some(storage) => Ok(storage.clone()),
+            None => {
+                let storage = Arc::new(SessionStorage::new().await?);
+                match SESSION_STORAGE.set(storage.clone()) {
+                    Ok(_) => Ok(storage),
+                    Err(_) => Ok(SESSION_STORAGE.get().unwrap().clone()),
+                }
+            }
+        }
+    }
+
+    pub async fn create_session(working_dir: PathBuf, description: String) -> Result<Session> {
+        let today = chrono::Utc::now().format("%Y%m%d").to_string();
+        let storage = Self::instance().await?;
+
+        let mut tx = storage.pool.begin().await?;
+
+        let max_idx = sqlx::query_scalar::<_, Option<i32>>(
+            "SELECT MAX(CAST(SUBSTR(id, 10) AS INTEGER)) FROM sessions WHERE id LIKE ?",
+        )
+        .bind(format!("{}_%", today))
+        .fetch_one(&mut *tx)
+        .await?
+        .unwrap_or(0);
+
+        let session_id = format!("{}_{}", today, max_idx + 1);
```

Comment:
Do we ever use the fact that the ids have these dates in them? If not, could just be an autoincrementing int in sqlite.